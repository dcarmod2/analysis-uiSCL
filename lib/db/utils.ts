import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
