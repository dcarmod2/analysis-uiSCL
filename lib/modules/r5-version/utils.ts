import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
