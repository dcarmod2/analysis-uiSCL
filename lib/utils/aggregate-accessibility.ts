import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
