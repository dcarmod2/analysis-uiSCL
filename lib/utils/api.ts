// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
