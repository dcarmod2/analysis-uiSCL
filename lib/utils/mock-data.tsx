import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'util'
import {constructor as XorShift} from 'xorshift'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN, TRAVEL_TIME_PERCENTILES} from '../constants'

import multi from './multi'

// Create a random number generator that is seeded
const seededRandom = new XorShift([1, 2, 3, 4])

export const makeMockStore = configureStore([multi, thunkMiddleware])

export const mockWithProvider = (children, mockData = mockStores.init) => {
  const store = makeMockStore(mockData)
  const wrapper = mount(<Provider store={store}>{children}</Provider>)

  return {
    snapshot: () => mountToJson(wrapper),
    store,
    wrapper
  }
}

export const mockModel = {
  accessGroup: 'accessGroupName',
  createdAt: '2017-11-08T12:30:50.133Z',
  createdBy: 'person@gmail.com',
  nonce: '12345',
  updatedAt: '2017-11-08T12:31:19.602Z',
  updatedBy: 'person@gmail.com'
}

export const mockSegmentGeom: GeoJSON.LineString = {
  type: 'LineString',
  coordinates: [
    [-122.0246, 36.9707],
    [-122.0279, 37.049],
    [-121.9799, 37.2299],
    [-121.9445, 37.324],
    [-121.936, 37.353],
    [-121.924, 37.365]
  ]
}

export const mockSegment: CL.ModificationSegment = {
  fromStopId: '1',
  toStopId: null,
  geometry: mockSegmentGeom,
  spacing: 3000,
  stopAtEnd: false,
  stopAtStart: false
}

export const mockStop1 = {
  autoCreated: false,
  bearing: 0,
  center: [36.9707, -122.0246],
  distanceFromStart: 0,
  index: 0,
  lat: 36.9707,
  lon: -122.0246,
  stopId: '1'
}

export const mockGtfsStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockGtfsStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockGtfsStops = [mockGtfsStop1, mockGtfsStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  name: 'Mock Pattern',
  pattern_id: 'mock-pattern-id',
  stops: mockGtfsStops,
  trips: [
    {
      duration: 1234,
      end_time: 13579,
      start_time: 12345,
      trip_headsign: 'To Downtown',
      trip_id: 'abcd',
      trip_short_name: 'The Express'
    }
  ]
}

export const mockRoute = {
  label: 'mock-route-label',
  patterns: [mockPattern],
  route_id: 'route1'
}

export const mockFeed = {
  checksum: 'MOCK',
  id: '1',
  routes: [mockRoute],
  stopsById: {
    1: mockGtfsStop1,
    2: mockGtfsStop2
  },
  stops: [mockGtfsStop1, mockGtfsStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  _id: 'timetable-id',
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  dwellTimes: [12],
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd',
  phaseFromTimetable: '',
  phaseFromStop: '',
  phaseAtStop: '',
  phaseSeconds: 300,
  exactTimes: false,
  modificationId: '1234',
  modificationName: 'Test Modification'
}

export const mockOpportunityDataset = {
  sourceName: 'lodes-data-2014',
  name: 'Workers total',
  _id: 'Workers_total'
}

const regionId = '1'

export const mockBundle = {
  ...mockModel,
  errorCode: '',
  name: 'Mock Bundle',
  _id: '1',
  centerLat: 38.90124,
  centerLon: -77.015615,
  feeds: [],
  regionId,
  serviceStart: '2018-09-27',
  serviceEnd: '2018-09-28',
  status: 'DONE',
  totalFeeds: 0
}

export const mockRegion = {
  ...mockModel,
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Region description',
  _id: regionId,
  name: 'Mock Region',
  opportunityDatasets: [mockOpportunityDataset],
  statusCode: 'DONE'
}

export const mockProject = {
  ...mockModel,
  bundleId: '1',
  _id: '1',
  name: 'Mock Project',
  regionId: mockRegion._id,
  variants: ['Default']
}

export const mockModification = {
  ...mockModel,
  bidirectional: false,
  description: 'Mock description',
  entries: [],
  feed: mockFeed.id,
  _id: '1234',
  name: 'Test Modification',
  routes: [mockRoute.route_id],
  projectId: mockProject._id,
  segments: [mockSegment],
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [true]
}

export const mockProfileRequest = {
  accessModes: 'WALK',
  bikeSafe: 1,
  bikeSlope: 1,
  bikeSpeed: 4.166666666666667,
  bikeTime: 1,
  bounds: mockRegion.bounds,
  bundleId: mockBundle._id,
  date: '2016-01-16',
  defaultExpanded: true,
  directModes: 'WALK',
  egressModes: 'WALK',
  fromLat: 12,
  fromLon: 34,
  fromTime: 25200,
  maxBikeTime: 20,
  maxRides: 4,
  maxTripDurationMinutes: 60,
  maxWalkTime: 20,
  monteCarloDraws: 200,
  name: 'Profile Request Test',
  projectId: mockProject._id,
  reachabilityThreshold: 0,
  regionId: mockRegion._id,
  toTime: 32400,
  transitModes: 'BUS,TRAM,RAIL,SUBWAY',
  travelTimePercentile: 50,
  variantIndex: 0,
  walkSpeed: 1.3888888888888888,
  workerVersion: 'v4.0.0'
}

export const mockRegionalAnalyses = [
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 0,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1,
    travelTimePercentiles: [50]
  },
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 1,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 50,
    travelTimePercentiles: [50]
  }
]

const GRID_SIZE = 300
const GRID_MAX = 2000
export const mockGrids = mockRegionalAnalyses.map((ra) => ({
  analysisId: ra._id,
  percentile: 50,
  cutoff: 60,
  zoom: 9,
  width: GRID_SIZE,
  height: GRID_SIZE,
  north: 20024,
  west: 20015,
  data: Array(GRID_SIZE * GRID_SIZE)
    .fill(0)
    .map(() => Math.floor(seededRandom.random() * GRID_MAX)),
  min: 0,
  max: GRID_MAX,
  contains: () => true
}))

export const mockGrid = mockGrids[0]

export const mockStores = {
  init: {
    analysis: {
      copyRequestSettings: true,
      isochroneFetchStatus: false,
      regional: {
        bounds: mockRegion.bounds
      },
      requestsSettings: [mockProfileRequest],
      resultsSettings: [],
      maxTripDurationMinutes: 60,
      travelTimePercentile: 50
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    opportunityDatasets: {
      active: mockOpportunityDataset._id,
      datasets: mockRegion.opportunityDatasets,
      uploadStatuses: []
    },
    queryString: {
      analysisId: mockRegionalAnalyses[0]._id,
      cutoff: 60,
      percentile: 50,
      modificationId: mockModification._id,
      projectId: mockProject._id,
      regionId: mockRegion._id
    },
    region: {
      aggregationAreas: [],
      bundles: [mockBundle],
      regions: [mockRegion]
    },
    regionalAnalyses: {
      analyses: mockRegionalAnalyses,
      grids: mockGrids
    },
    project: {
      activeModificationId: mockModification._id,
      currentProject: mockProject,
      feeds: [mockFeed],
      modifications: [mockModification],
      projects: [mockProject]
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  modificationId: 'Add Trip Pattern',
  title: "Error while applying the modification entitled 'Add Trip Pattern'.",
  messages: ['Headway is not greater than zero.']
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  depth: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

/**
 * This is a function because responseToSurface is destructive and can lead to
 * test leakage if we don't have a brand-new surface each time.
 */
export const createMockSurfaceData = () => {
  const {width, height, depth, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize =
    9 * 4 + // header
    width * height * depth * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * depth)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = depth

  // phew, header is done
  for (let z = 0, position = 9; z < depth; z++) {
    let prev = 0
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++, position++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[position] = pixelVal - prev // values are delta-coded within each depth
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(
    buffer,
    buffer.byteLength - metadataEncoded.byteLength,
    metadataEncoded.length
  )
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

/**
 * Return the value stored in sample z of pixel x, y of the mock surface
 */
export function getMockSurfaceValue(x, y, z) {
  // This used to simply return z, 2 * z and 3 * z when we had an array of
  // [5, 10, 15 . . . 95]th percentiles. However, when we switched to
  // requesting specific percentiles that stopped working. This translates the
  // position in the access grid into a percentile, then subtracts five and
  // divides by five to get the position in the old array, so that tests
  // written assuming the old structure continue to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return (2 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  if (x < 2 && y >= 2) return (3 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  return 0
}

const opportunityDatasetGridData = new Int32Array(
  mockSurfaceParameters.width * mockSurfaceParameters.height
)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) {
  opportunityDatasetGridData[i] = 1
}
for (
  let i = 3 * mockSurfaceParameters.width;
  i < opportunityDatasetGridData.length;
  i++
) {
  opportunityDatasetGridData[i] = 2
}

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: opportunityDatasetGridData,
  getValue(x, y) {
    return opportunityDatasetGridData[y * mockSurfaceParameters.width + x]
  }
}

export const releaseVersions = ['v3.0.0', 'v2.2.0', 'v2.0.0']
export const allVersions = [
  'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0'
]
