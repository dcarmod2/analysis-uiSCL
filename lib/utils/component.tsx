import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
