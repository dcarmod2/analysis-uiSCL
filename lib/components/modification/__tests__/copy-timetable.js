import clone from 'lodash/cloneDeep'

import {testComponent} from 'lib/utils/component'
import {mockModification, mockSegment} from 'lib/utils/mock-data'
import CopyTimetable from '../copy-timetable'

const mockRegionId = 'r1'

const mockProject = {
  _id: 'p1',
  name: 'mock project'
}

// response only includes regions, projects and modifications that
// all have at least one timetable
const mockResponse = [
  {
    projects: [
      {
        ...mockProject,
        modifications: [
          {
            timetables: [
              {
                _id: 't1',
                monday: true,
                tuesday: true,
                wednesday: true,
                thursday: true,
                friday: true,
                saturday: false,
                sunday: false,
                name: 'M-F',
                startTime: 25200,
                endTime: 79200,
                headwaySecs: 600,
                exactTimes: false,
                phaseAtStop: null,
                phaseFromTimetable: null,
                phaseFromStop: null,
                phaseSeconds: 0,
                dwellTime: 0,
                segmentSpeeds: [16],
                dwellTimes: []
              }
            ],
            name: 'mock modification',
            _id: 'm1',
            segments: [mockSegment]
          }
        ]
      }
    ],
    name: 'mock region',
    _id: mockRegionId
  }
]

function timeoutPromise() {
  return new Promise((resolve) => {
    setTimeout(resolve, 5)
  })
}

describe('Component > Modification > CopyTimetable', () => {
  it('renders correctly', async () => {
    const create = jest.fn()
    const wrapper = testComponent(CopyTimetable, {
      create,
      intoModification: mockModification
    }).shallow()

    // Snapshot the loaded state
    expect(wrapper).toMatchSnapshot()

    // Click the button
    wrapper.find('Button').simulate('click')

    // Modal should be open
    expect(wrapper).toMatchSnapshot()

    // Unmount wihtout errors
    wrapper.unmount()
  })

  // if rendering a region without any timetables, show the first
  // timetable that is available
  it('renders with a region without timetables', async () => {
    const wrapper = testComponent(CopyTimetable, {
      create: jest.fn(),
      intoModification: mockModification
    }).shallow()

    // wait for load to finish
    await timeoutPromise()

    // snapshot loaded state
    expect(wrapper).toMatchSnapshot()

    wrapper.unmount()
  })

  // a test case for when no timetables are available to copy
  it('renders when no timetables exist in database', async () => {
    const wrapper = testComponent(CopyTimetable, {
      create: jest.fn(),
      intoModification: mockModification
    }).shallow()

    // Wait for load to finish
    await timeoutPromise()

    // snapshot loaded state
    expect(wrapper).toMatchSnapshot()

    // unmount
    wrapper.unmount()
  })

  describe('segment warnings', () => {
    it('shows a warning when current modification has no segments', async () => {
      const wrapper = testComponent(CopyTimetable, {
        create: jest.fn(),
        intoModification: {
          ...mockModification,
          segments: []
        }
      }).shallow()

      // wait for load to finish
      await timeoutPromise()

      // snapshot loaded state
      expect(wrapper).toMatchSnapshot()

      wrapper.unmount()
    })

    it('shows a warning when modification of timetable being copied has no segments', async () => {
      const serverResponse = clone(mockResponse)
      serverResponse[0].projects[0].modifications[0].segments = []
      serverResponse[0].projects[0].modifications[0].timetables[0].segmentSpeeds = []

      const wrapper = testComponent(CopyTimetable, {
        create: jest.fn(),
        intoModification: mockModification
      }).shallow()

      // wait for load to finish
      await timeoutPromise()

      // snapshot loaded state
      expect(wrapper).toMatchSnapshot()

      wrapper.unmount()
    })

    it('shows a warning when modification of timetable being copied has more segments', async () => {
      const serverResponse = clone(mockResponse)
      serverResponse[0].projects[0].modifications[0].segments.push(mockSegment)
      serverResponse[0].projects[0].modifications[0].timetables[0].segmentSpeeds.push(
        17
      )

      const wrapper = testComponent(CopyTimetable, {
        create: jest.fn(),
        intoModification: mockModification
      }).shallow()

      // wait for load to finish
      await timeoutPromise()

      // snapshot loaded state
      expect(wrapper).toMatchSnapshot()

      wrapper.unmount()
    })

    it('shows a warning when modification of timetable being copied has less segments', async () => {
      const wrapper = testComponent(CopyTimetable, {
        create: jest.fn(),
        intoModification: {
          ...mockModification,
          segments: [...mockModification.segments, mockSegment]
        }
      }).shallow()

      // wait for load to finish
      await timeoutPromise()

      // snapshot loaded state
      expect(wrapper).toMatchSnapshot()

      wrapper.unmount()
    })
  })
})
