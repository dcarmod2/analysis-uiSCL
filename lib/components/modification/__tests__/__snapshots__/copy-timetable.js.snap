// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > Modification > CopyTimetable renders correctly 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable renders correctly 2`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
  <Modal
    closeOnOverlayClick={false}
    isOpen={true}
    onClose={[Function]}
  >
    <ModalOverlay />
    <ModalContent>
      <ModalHeader>
        Copy timetable
      </ModalHeader>
      <ModalCloseButton />
      <CopyTimetableContent
        create={[MockFunction]}
        intoModification={
          Object {
            "_id": "1234",
            "accessGroup": "accessGroupName",
            "bidirectional": false,
            "createdAt": "2017-11-08T12:30:50.133Z",
            "createdBy": "person@gmail.com",
            "description": "Mock description",
            "entries": Array [],
            "feed": "1",
            "name": "Test Modification",
            "nonce": "12345",
            "projectId": "1",
            "routes": Array [
              "route1",
            ],
            "segments": Array [
              Object {
                "fromStopId": "1",
                "geometry": Object {
                  "coordinates": Array [
                    Array [
                      -122.0246,
                      36.9707,
                    ],
                    Array [
                      -122.0279,
                      37.049,
                    ],
                    Array [
                      -121.9799,
                      37.2299,
                    ],
                    Array [
                      -121.9445,
                      37.324,
                    ],
                    Array [
                      -121.936,
                      37.353,
                    ],
                    Array [
                      -121.924,
                      37.365,
                    ],
                  ],
                  "type": "LineString",
                },
                "spacing": 3000,
                "stopAtEnd": false,
                "stopAtStart": false,
                "toStopId": null,
              },
            ],
            "timetables": Array [
              Object {
                "_id": "timetable-id",
                "dwellTime": 10,
                "dwellTimes": Array [
                  12,
                ],
                "endTime": 57600,
                "exactTimes": false,
                "friday": true,
                "headwaySecs": 900,
                "modificationId": "1234",
                "modificationName": "Test Modification",
                "monday": true,
                "name": "Test timetable",
                "patternTrips": Array [
                  "abcd",
                ],
                "phaseAtStop": "",
                "phaseFromStop": "",
                "phaseFromTimetable": "",
                "phaseSeconds": 300,
                "saturday": false,
                "segmentSpeeds": Array [],
                "sourceTrip": "abcd",
                "startTime": 28800,
                "sunday": false,
                "thursday": true,
                "tuesday": true,
                "wednesday": true,
              },
            ],
            "trips": Array [
              "abcd",
            ],
            "type": "add-trip-pattern",
            "updatedAt": "2017-11-08T12:31:19.602Z",
            "updatedBy": "person@gmail.com",
            "variants": Array [
              true,
            ],
          }
        }
        onClose={[Function]}
      />
    </ModalContent>
  </Modal>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable renders when no timetables exist in database 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable renders with a region without timetables 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable segment warnings shows a warning when current modification has no segments 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable segment warnings shows a warning when modification of timetable being copied has less segments 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable segment warnings shows a warning when modification of timetable being copied has more segments 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;

exports[`Component > Modification > CopyTimetable segment warnings shows a warning when modification of timetable being copied has no segments 1`] = `
<Fragment>
  <Button
    isFullWidth={true}
    leftIcon="copy"
    onClick={[Function]}
    variantColor="green"
  >
    Copy existing timetable
  </Button>
</Fragment>
`;
